staging/src/k8s.io/kube-aggregator/pkg/apiserver/handler_discovery.go

```diff
@@ -0,0 +1,575 @@
+	// implementation?
+	port := int32(443)
+	if service.Port != nil {
+		port = *service.Port
+	}
+
+	return serviceKey{
+		Name:      service.Name,
+		Namespace: service.Namespace,
+		Port:      port,
+	}
+}
+
+type cachedResult struct {
+	// Currently cached discovery document for this service
+	// Map from group name to version name to
+	discovery map[metav1.GroupVersion]apidiscoveryv2beta1.APIVersionDiscovery
+
+	// ETag hash of the cached discoveryDocument
+	etag string
+
+	// Guaranteed to be a time less than the time the server responded with the
+	// discovery data.
+	lastUpdated time.Time
+
+	// results are stale if server was attempted to be contacted and it failed
+	// if lastUpdated is beyond a certain threshold, and stale is true, the result
+	// will be purged
+	stale bool
+}
+
+// Information about a specific APIService/GroupVersion
+type groupVersionInfo struct {
+	// Date this APIService was marked dirty.
+	// Guaranteed to be a time greater than the most recent time the APIService
+	// was known to be modified.
+	//
+	// Used for request deduplication to ensure the data used to reconcile each
+	// apiservice was retrieved after the time of the APIService change:
+	// real_apiservice_change_time < groupVersionInfo.lastMarkedDirty < cachedResult.lastUpdated < real_document_fresh_time
+	//
+	// This ensures that if the apiservice was changed after the last cached entry
+	// was stored, the discovery document will always be re-fetched.
+	lastMarkedDirty time.Time
+
+	// Last time sync function was run for this GV.
+	lastReconciled time.Time
+
+	// ServiceReference of this GroupVersion. This identifies the Service which
+	// describes how to contact the server responsible for this GroupVersion.
+	service serviceKey
+
+	// groupPriority describes the priority of the APIService for sorting
+	groupPriority int
+
+	// Method for contacting the service
+	handler http.Handler
+}
+
+var _ DiscoveryAggregationController = &discoveryManager{}
+
+func NewDiscoveryManager(
+	target discoveryendpoint.ResourceManager,
+) DiscoveryAggregationController {
+	return &discoveryManager{
+		mergedDiscoveryHandler: target,
+		apiServices:            make(map[string]groupVersionInfo),
+		cachedResults:          make(map[serviceKey]cachedResult),
+		dirtyAPIServiceQueue:   workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "discovery-manager"),
+	}
+}
+
+// Returns discovery data for the given apiservice.
+// Caches the result.
+// Returns the cached result if it is retrieved after the apiservice was last
+// marked dirty
+func (dm *discoveryManager) fetchFreshDiscoveryForService(gv metav1.GroupVersion, info groupVersionInfo) (*cachedResult, error) {
+	// Lookup last cached result for this apiservice's service.
+	cached, exists := dm.getCacheEntryForService(info.service)
+
+	// If entry exists and was updated after the given time, just stop now
+	if exists && !cached.stale && cached.lastUpdated.After(info.lastMarkedDirty) {
```

Comment:
Time is tricky.  What are our vector clock options?  When time moves backwards, this can produce subpar outcomes.

Reply:
Can you elaborate on this? Time is local to each apiserver, and is never shared or persisted. What kind of error do you see occurring?

Reply:
according to https://pkg.go.dev/time#hdr-Monotonic_Clocks

> The Time returned by time.Now contains a monotonic clock reading. 
> If Times t and u both contain monotonic clock readings, the operations t.After(u), t.Before(u), t.Equal(u), and t.Sub(u) are carried out using the monotonic clock readings alone, ignoring the wall clock readings. If either t or u contains no monotonic clock reading, these operations fall back to using the wall clock readings.

@deads2k given the above, it seems safe to me to use `t.After` to compare two `time.Now` readings?

Reply:
> @deads2k given the above, it seems safe to me to use t.After to compare two time.Now readings. is there something else to watch out for?

When NTP resets a clock backwards by several hours or even days, these times can be in the future so far they don't expire for a very long time.  We see time moving backwards in customer clusters running on their own hardware.   Could this be handled as atomic ints instead?  If I've properly interpreted the "happens-before" I think the answer is yes.

Reply:
From the link above:
> Operating systems provide both a “wall clock,” which is subject to changes for clock synchronization, and a “monotonic clock,” which is not.

I think what Alex is saying is that the monotonic clock won't be affected by changes to system time. It's using a separate clock (which, I agree, is very confusing in this case ...)

Reply:
If you want to be extra safe, you could also enter the `if` if the dirty time is in the future? I am not sure how far I trust go's monotonic clock

Reply:
If this is an in memory object I'm not sure it's super likely for kubectl to hit this condition, since it's not long-running. I'd worry more when (if?) we make controllers use this. Maybe put a TODO here for more investigation and don't worry about it for now?

Reply:
Wait this is the server PR, nvm

Reply:
I didn't realize golang did that. How convenient, that satisfies my concern.